### PR TITLE
patched IBE hashToCurve with ietf hashToField impl

### PR DESCRIPTION
**Pull request by CluEleSsUK**
- pulled some pieces out of noble
- wrote an implementation of `hashToField` using https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16#section-5.2
- created a patched `hashToCurve`

### CODE REVIEW

**Comment by CluEleSsUK** [comment]
Took a first crack at this - only 3 tests failing including the go compatibility ones

**Comment by nikkolasg** [comment]
What's the reason of having our own implementation of that ?
Noble seems to support the RFC too. is the difference in version the reason ? (12 for Noble vs 16 for this PR).

**Comment by CluEleSsUK** [comment]
In retrospect, I perhaps should have read the feedback from the audit more clearly - it looks like only the rejection sampling in `h3` needs changed, not the whole hashToField x) I took the recommendation at the end to implement the ietf one a bit too literally oops

**Comment by AnomalRoil** [comment]
Yeah, in theory Noble should implement what we need already.

**Comment by nikkolasg** [comment]
Can someone link then the reason and/or issue raised in the audit (or paste a summary etc), in this PR or issue here ? It'd be nice that this PR contains everything one needs to know to understand the reasons, impacts etc.

**Comment by AnomalRoil** [comment]
Our H3 function is vulnerable to collision. But I'd argue it's not a problem since we do key-wrapping and keys have fixed sizes, but TBH, using HashToField and HashToField from the RFC is probably better

**Review by AnomalRoil — CHANGES_REQUESTED** [review]
Could we switch to the Noble implementations? If you want to wrap them in your own function to change their API, that's fine.
Also let's keep the tests.

File: src/crypto/hash-to-curve.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
+import {CURVE, Fp, Fp2, PointG2} from "@noble/bls12-381"
+import {hexToBytes} from "@noble/hashes/utils"
+import {hashToField} from "./hash-to-field"
+
+// this patched `hashToCurve` uses a different implementation of `hashToField` than noble
+// see: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16
+export async function hashToCurve(input: Uint8Array | string): Promise<PointG2> {
```
[the remark this answers is not in the document]
Comment (CluEleSsUK) [reply src/crypto/hash-to-curve.ts:7]: we were actually using them already internally - I think we can just update the package and then change our h3 to comply

**Comment by CluEleSsUK** [comment]
Closing this because I got the wrong end of the stick (was a good learning exercise though ;p)

**Pull request closed without merging**